eth/handler.go

```diff
@@ -288,22 +292,69 @@ func newHandler(config *handlerConfig) (*handler, error) {
 	return h, nil
 }
 
+// protoTracker tracks the number of active protocol handlers.
+func (h *handler) protoTracker() {
+	defer h.wg.Done()
+	var (
+		active int
+		done   = make(chan struct{})
+	)
+	// Start goroutine to track active protocol handlers.
+	go func() {
+		for {
+			select {
+			case <-h.handlerStartCh:
+				active--
+			case <-h.handlerDoneCh:
+				active++
+			case <-done:
+				return
+			}
+		}
+	}()
+	<-h.quitSync
+
+	// After quit is initiated, poll until all handlers are closed.
+	for range time.Tick(time.Millisecond * 50) {
+		if active == 0 {
+			break
+		}
+	}
+	done <- struct{}{}
+}
+
+// incHandlers signals to increment the number of active handlers if not
+// quitting.
+func (h *handler) incHandlers() bool {
+	select {
+	case <-h.quitSync:
+		return false
+	default:
+		h.handlerStartCh <- struct{}{}
+		return true
+	}
+}
```

Comment:
The logics here are a bit devious. So this first checks if quitsync is happening. If not, it waits, blockingly, to send something on `handlerStartCh`. 
I'm suspecting that there's a potential deadlock here...? But I'm not sure, I don't have the full picture of the statemachine internalized

Reply:
A more 'typical' switch wold be something like this
```
select { 
  case <- h.quitSync: 
       return false
   case h.handlerStartCh <- struct{}{}:
       return true
  } 
```

Reply:
Basically it is verifying that we aren't actively quitting. This is the logic in [handlePeerEvent`](https://github.com/ethereum/go-ethereum/blob/c2db667c8f824ec0d025cbde95e6aaad67fc10ba/eth/sync.go#L92-L99) which was previously at the beginning of the `runXXXXPeer`.

I guess the situation I'm not super sure about is say we get to line 333 here and then somehow the runtime decides to pause the thread long enough so that `protoTracker` is torn down completely, then it resumes at 333 and there is nothing listening to the channel anymore and it blocks forever.

Reply:
Ah okay I see, that makes more sense without the default.

Reply:
> I guess the situation ...

Yeah it was exactly that which I meant